### ISSUE TITLE
New key is created every time query is made

### ISSUE THREAD

**Issue by robert197** [noripyt/django-cachalot#161]
## What happened?
Everytime a query is made with specific query on my system, a new key in redis is created

## What should've happened instead?
I guess only on the first time time a key should be created and should stay until its invalidated?
I think there should be one key per query and not everytime a new key.

## Steps to reproduce
I use one query.
Then i query for keys with "KEYS *" in redis CLI and see for e.g. 10 keys.
Execute the same query again.
After this i see 11 keys

[//]: # (Include host configuration: OS, Django version, database, etc.)

**Comment by Andrew-Chen-Wang** [comment]
Hi @robert197, 

Take a look at how we generate cache keys:

https://github.com/noripyt/django-cachalot/blob/6781b9b98a1bb7502801772f16ac83f32ca90b04/cachalot/utils.py#L58-L75

Are you sure the inputs are the same? Every single query, even with the slightest difference, will generate a new cache key with a certain expiration lifetime (Redis will do the passive expired cache clean).

**Comment by robert197** [comment]
The query i tested this with is very simple.

```
@login_required
def resolve_kantons(self, info, **kwargs):
    return Kanton.objects.all() 
```
Looking on the code of key generation. It should work.
Im using graphql and graphene and as api.

Do you think it's because of graphene?

**Comment by Andrew-Chen-Wang** [comment]
It may be because of graphene. Do you mind posting the Redis keys? I've ran a successful test case like with atomic transactions (for only the first one):
```
    def test_blah(self):
        with self.assertNumQueries(1):
            Test.objects.create()
        with self.assertNumQueries(1 if self.is_sqlite else 0):  # Starts with BEGIN but no real queries for sqlite
            with transaction.atomic():
                Test.objects.all()
        with self.assertNumQueries(1 if self.is_sqlite else 0):  # Starts with BEGIN but no real queries for sqlite
            with transaction.atomic():
                Test.objects.all()
        self.assert_query_cached(Test.objects.all())

    def test_blah1(self):
        with self.assertNumQueries(1):
            Test.objects.create()
        with self.assertNumQueries(0):
            Test.objects.all()
        with self.assertNumQueries(0):
            Test.objects.all()
        self.assert_query_cached(Test.objects.all())
```
Try removing the `@login_required` decorator and see if it still happens.

**Comment by robert197** [comment]
Output of `qs.query` is following:

``` sql
SELECT "orders_and_offers_kanton"."sort_order", "orders_and_offers_kanton"."id", "orders_and_offers_kanton"."name"
FROM "orders_and_offers_kanton"
ORDER BY "orders_and_offers_kanton"."sort_order" ASC
```

After this i removed `print(qs.query)`

Key checks:

- First time query generates 6 keys (output of `KEYS *` in *redis-cli*):

``` bash
127.0.0.1:6379> keys *
1) ":1:e834c2cb34d13799de259ef4bfbec225df6c1a15"
2) ":1:a2cec6efb2cb51d9fcdc73ec65d0d02848d3205f"
3) ":1:8a85a2a2c4fcae28335c9d19fcd219ec8bb9f3bf"
4) ":1:65a4933e20fe156362340b4953576c5a57f25864"
5) ":1:aa9bee870d4c3f171f94536813db534cc345e3b7"
6) ":1:f6e1a4d48790db63f8cb8c5653192a74ca00d34a"
```


- After executing the same query second time i got one key more:

``` bash
127.0.0.1:6379> keys *
1) ":1:94bd9c8ab7b620f7059315c311186667a99ee83c"
2) ":1:e834c2cb34d13799de259ef4bfbec225df6c1a15"
3) ":1:a2cec6efb2cb51d9fcdc73ec65d0d02848d3205f"
4) ":1:8a85a2a2c4fcae28335c9d19fcd219ec8bb9f3bf"
5) ":1:65a4933e20fe156362340b4953576c5a57f25864"
6) ":1:aa9bee870d4c3f171f94536813db534cc345e3b7"
7) ":1:f6e1a4d48790db63f8cb8c5653192a74ca00d34a"
```


And every time i execute the same query i get one more key...

BTW:
I removed `@login_required` for this tests.

**Comment by Andrew-Chen-Wang** [comment]
Try this:
- Run that function three times instead of the usual two you've been doing in this issue.
- Use django-debug-toolbar and output the sql, caches, and caches from cachalot's panel.
- In your `resolve_kanton` function, perform that query three times. Output the number of queries. Also try it with the query executing twice.

---
Ok the fun details:

Huh ok this is an interesting problem... I think it could be graphene. Did you try just performing the query in a regular Django view (we have a cachalot panels for django-debug-toolbar)? If you don't mind, please use Django-debug-toolbar and show me any SQL statements and caches that the package records inside your view (with Django-cachalot's cache panel included, too!). Double check there are no external AJAX calls being made. Make sure there's only one database being used and not two.

Finally, in your original code posted, check to see if the cache is created if you run the query again (i.e. perform it thrice):

```
def resolve_kantons(self, info, **kwargs):
    Kanton.objects.all()
    Kanton.objects.all() 
    return Kanton.objects.all() 
```

then check if multiple cache keys are generated. If only one cache is generated every time you call that function `resolve_kantons`, then it's a problem in your application. If three cache keys are made in that function on the first call, then it's a cachalot problem.

The one thing I find weird is that Redis generates 6 keys... 6? This leads me to believe something else is being done that generates that extra key rather than your application specifically.

**TL;DR:** Before I get into my detailed analysis, just know that I have zero clue where that extra cache key is coming from. It must be coming from some internal system or third-party package that's creating that extra key.

Ok, it's kinda late where I'm at, but this is how I'm trying to re-produce your error:

models.py
```
from django.db import models


class Blah(models.Model):
    sort_order = models.IntegerField(default=1)
    id = models.AutoField(primary_key=True)
    name = models.CharField(default="asdf", max_length=100)

    class Meta:
        db_table = "orders_and_offers_kanton"
        ordering = ["sort_order"]
```

views.py
```
from django.shortcuts import render
from hashlib import sha1
from .models import Blah
from django.db.models.sql.compiler import SQLCompiler
from django.db import connection


def index(request):
    q = (Blah.objects.all()).query
    compiler = SQLCompiler(query=q, connection=connection, using="default")
    sql, params = compiler.as_sql()
    print(sql)
    print(params)
    cache_key = '%s:%s:%s' % (compiler.using, sql,
                              [str(p) for p in params])
    print(cache_key)
    print(sha1(cache_key.encode('utf-8')).hexdigest())
    return render(request, "index.html")
```

Output:
```
SELECT "orders_and_offers_kanton"."sort_order", "orders_and_offers_kanton"."id", "orders_and_offers_kanton"."name" FROM "orders_and_offers_kanton" ORDER BY "orders_and_offers_kanton"."sort_order" ASC
()
default:SELECT "orders_and_offers_kanton"."sort_order", "orders_and_offers_kanton"."id", "orders_and_offers_kanton"."name" FROM "orders_and_offers_kanton" ORDER BY "orders_and_offers_kanton"."sort_order" ASC:[]
e834c2cb34d13799de259ef4bfbec225df6c1a15
```

So I've managed to match with the hex `e834c2cb34d13799de259ef4bfbec225df6c1a15`. But I can't get that extra key.

Permutations of your model:

```
>>> list(itertools.permutations(["name", "id", "sort_order"]))
[('name', 'sort_order', 'id'), ('name', 'id', 'sort_order'), ('sort_order', 'name', 'id'), ('sort_order', 'id', 'name'), ('id', 'name', 'sort_order'), ('id', 'sort_order', 'name')]
```

And then the hex that followed each (when you reorder them in models.py you change how they are ordered in the generated sql):
```
('name', 'sort_order', 'id')  480bd56627ad0ccb90a8aadb3d5100329d99ded9
('name', 'id', 'sort_order')  93a98dc72b2ba6b88438610571e610581bc63ddf
('sort_order', 'name', 'id')  5d61c50032617b6ea451be2d66ee2ace244423a8
('sort_order', 'id', 'name')  e834c2cb34d13799de259ef4bfbec225df6c1a15
('id', 'name', 'sort_order')  9799726a1c8586fd92a6b5ce37ae5c7a7843dc7d
('id', 'sort_order', 'name')  a2d24acbc8403b5becea898b161bf7097e414185
```

Finally without ordering:
```
('name', 'sort_order', 'id')  23948dfc9612ceb298af47515a81460cea09b983
('name', 'id', 'sort_order')  a7f57149bec0ea8490d30c4a37b6bf956702d5c3
('sort_order', 'name', 'id')  ed6a92f7c73b48a6cf2c483fbe996f155f6ae015
('sort_order', 'id', 'name')  f22ebce57f762b21e6c7a51ce82059f051334271
('id', 'name', 'sort_order')  a7f57149bec0ea8490d30c4a37b6bf956702d5c3
('id', 'sort_order', 'name')  b9925fd1ddd12126478627af0e3b9709d1338af2
```

I couldn't match that extra key (ie `94bd9c8ab7b620f7059315c311186667a99ee83c`) with any of the permutations.

Final nail in the coffin is the table cache key generation:
```
>>> from hashlib import sha1
>>> cache_key = "%s:%s" % ("default", "orders_and_offers_kanton")
>>> cache_key
'default:orders_and_offers_kanton'
>>> sha1(cache_key.encode("utf-8")).hexdigest()
'65a4933e20fe156362340b4953576c5a57f25864'
```

**Comment by Andrew-Chen-Wang** [comment]
Closing for lack of response and I can't reproduce the error.
